### PR TITLE
Add german TIN validator

### PR DESCRIPTION
**Pull request by bofalke**
This PR will implements the Validor for German Tax Identification Number (TIN) mentioned in #47

### CODE REVIEW

File: src/Core/Europe/Germany/GermanIdValidator.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+namespace Reducktion\Socrates\Core\Europe\Germany;
+
+use Reducktion\Socrates\Contracts\IdValidator;
+use Reducktion\Socrates\Exceptions\InvalidLengthException;
+
+class GermanIdValidator implements IdValidator
+{
+    public function validate(string $id): bool
+    {
+        $idlength = strLen($id);
```
Comment (AlexOlival) [inline src/Core/Europe/Germany/GermanIdValidator.php:12]: I've never seen this function call, does this work? 🤔 
Even if it does, could you change to the common syntax for consistency please?
https://www.php.net/manual/en/function.strlen.php

File: src/Core/Europe/Germany/GermanIdValidator.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        if (11 !== $idlength) {
```
Comment (AlexOlival) [inline src/Core/Europe/Germany/GermanIdValidator.php:13]: Yoda conditions are great, but we haven't been using them. For the sake of consistency let's swap both terms of the statement. Also, please rename the variable to `$idLength`

File: src/Core/Europe/Germany/GermanIdValidator.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            throw new InvalidLengthException('German TIN', '11', $idlength);
+        }
+
+        if ($this->isFirstDigitZero($id)) {
+            return false;
+        }
+
+        if (!$this->isDigitsHasCorrectStats($id)) {
+            return false;
+        }
```
Comment (AlexOlival) [inline src/Core/Europe/Germany/GermanIdValidator.php:23]: You can keep these, but I think it's fine to inline these operations.
In case you prefer to keep the functions, please change `isDigitsHasCorrectStats` to `checkDigitsHaveCorrectStats` or something more grammatically correct.

I think inlining is still preferable 😛

File: tests/Feature/Europe/GermanyTest.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            '48954371207',
+            '55492670836'
+        ];
+
+        $this->invalidIds = [
+            '01234567812',
+            '81872495631',
+            '48954371206',
+            '55492670834'
+        ];
```
Comment (AlexOlival) [inline tests/Feature/Europe/GermanyTest.php:31]: Can we get at least five test cases for each, please?

**Review by AlexOlival — CHANGES_REQUESTED** [review]
Thank you for your work in this!
I left some comments to be addressed.

File: tests/Feature/Europe/GermanyTest.php (hunk 4 above)
Comment (bofalke) [reply tests/Feature/Europe/GermanyTest.php]: Yes sure, i found only those three in some documentation thats why i chose them. I made two more up and hope they do not belong to anyone..

File: src/Core/Europe/Germany/GermanIdValidator.php (hunk 3 above)
Comment (bofalke) [reply src/Core/Europe/Germany/GermanIdValidator.php]: Inlined everything :+1:

File: src/Core/Europe/Germany/GermanIdValidator.php (hunk 1 above)
Comment (bofalke) [reply src/Core/Europe/Germany/GermanIdValidator.php]: It works but was not intended, good catch

File: src/Core/Europe/Germany/GermanIdValidator.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        if ((int) $id[0] === 0) {
+            return false;
+        }
+
+        // The first 10 numbers of the German TIN need to have 8 or 9 different numbers
+        $digits = str_split($id);
+        array_pop($digits);
+        $countDigits = count(array_unique($digits));
+
+        if ((9 !== $countDigits && 8 !== $countDigits)) {
```
Comment (AlexOlival) [inline src/Core/Europe/Germany/GermanIdValidator.php:27]: Invert these yoda conditions too, please

File: src/Core/Europe/Germany/GermanIdValidator.php [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            return false;
+        }
+
+        // Checksum calculation
+        $product = 10;
+        for ($i = 0; $i <= 9; ++$i) {
+            $sum = ((int) $id[$i] + $product) % 10;
+            if (0 === $sum) {
```
Comment (AlexOlival) [inline src/Core/Europe/Germany/GermanIdValidator.php:35]: Ditto as above

File: src/Core/Europe/Germany/GermanIdValidator.php [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+                $sum = 10;
+            }
+            $product = ($sum * 2) % 11;
+        }
+        $checksum = 11 - $product;
+        if (10 === $checksum) {
```
Comment (AlexOlival) [inline src/Core/Europe/Germany/GermanIdValidator.php:41]: Ditto

**Review by AlexOlival — APPROVED** [review]

File: tests/Feature/Europe/GermanyTest.php [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+        foreach ($this->invalidIds as $invalidId) {
+            $this->assertFalse(
+                Socrates::validateId($invalidId, 'DE')
+            );
+        }
+
+        $this->expectException(InvalidLengthException::class);
+
+        Socrates::validateId('0123456789', 'GR');
```
Comment (AlexOlival) [inline tests/Feature/Europe/GermanyTest.php:60]: Not the correct country code

**Review by AlexOlival — CHANGES_REQUESTED** [review]
Some more small tweaks

**Review by AlexOlival — APPROVED** [review]
We're there! Thank you!
@JoaoFSCruz please come take a looksie.

Also, one last request - but I can take care of that later if you don't feel like it.
Could you change the readme to reflect the fact that we're no longer not supporting Germany?

**Review by JoaoFSCruz — CHANGES_REQUESTED** [review]
Thank you for your contribution! Nicely done ✔️ 

Just a couple of small things. 
- Please include the source of the algorithm so it's easier to maintain the code
- Update the README.md and Countries.md

**Comment by bofalke** [comment]
For the source of the algorithm should i link the original repo that i found? Other documentation is mostly in German, so i do not know if that's of any help?

**Comment by AlexOlival** [comment]
> For the source of the algorithm should i link the original repo that i found? Other documentation is mostly in German, so i do not know if that's of any help?

It's fine if the documentation is in another language. Any (correct) source will do :pray: 
If the documentation is enough, then we don't need to link to any other implementations imo

File: COUNTRIES.md [hunk 9]
```diff
@@ -42,3 +42,4 @@
 | Ukraine 🇺🇦                |      UA      | :heavy_check_mark: | :heavy_check_mark: |
 | United Kingdom 🇬🇧         |      GB      | :heavy_check_mark: | :x:                |
 | Uruguay 🇺🇾                |      UY      | :heavy_check_mark: | :x:                |
+| Germany 🇩🇪                |      DE      | :heavy_check_mark: | :x:                |
```
Comment (AlexOlival) [inline COUNTRIES.md:45]: We need this list to be ordered alphabetically by the country name please

File: README.md [hunk 10]
```diff
@@ -104,7 +104,7 @@ echo $citizen->getPlaceOfBirth(); // null
 
 [Here](COUNTRIES.md) you can see the full list of supported countries and whether they support data extraction.
 
-Four european countries are currently unsupported: Austria 🇦🇹, Belarus 🇧🇾, Cyprus 🇨🇾 and Germany 🇩🇪. This is because we could not find a reliable source for the algorithm, if at all. Help would be appreciated to get these countries supported.
+Four european countries are currently unsupported: Austria 🇦🇹, Belarus 🇧🇾 and Cyprus 🇨🇾. This is because we could not find a reliable source for the algorithm, if at all. Help would be appreciated to get these countries supported.
```
Comment (AlexOlival) [inline README.md:107]: Three* :grin:

**Review by AlexOlival — CHANGES_REQUESTED** [review]
Two small corrections on the markdown changes

File: README.md (hunk 10 above)
Comment (bofalke) [reply README.md:107]: Ooops :D

**Comment by bofalke** [comment]
There is an official document by ELSTER (that is the german institute to make your tax declarations online). They have a downloadable pdf where the details are explained in Chapter 2.2.

https://download.elster.de/download/schnittstellen/Pruefung_der_Steuer_und_Steueridentifikatsnummer.pdf

Should I use the download URL  to this PDF or this a problem since download urls tend to disappear sooner or later?

**Comment by AlexOlival** [comment]
> There is an official document by ELSTER (that is the german institute to make your tax declarations online). They have a downloadable pdf where the details are explained in Chapter 2.2.
> 
> https://download.elster.de/download/schnittstellen/Pruefung_der_Steuer_und_Steueridentifikatsnummer.pdf
> 
> Should I use the download URL to this PDF or this a problem since download urls tend to disappear sooner or later?

It's fine to link to the PDF
That's a risk we're taking with all links - but at least one can derive some information from the URL itself and try to find another source :+1:

**Comment by bofalke** [comment]
>It's fine to link to the PDF
That's a risk we're taking with all links - but at least one can derive some information from the URL itself and try to find another source :+1: 

Allright, i will link it. 

I just saw that they have some example values as well for the TIN in the documentation, maybe i should use those in our tests as well? 
There is a disclaimer below saying something like:
"The ID was taken from the concept of the Federal Information Technology Center and serves
solely for the purpose of illustrating the implementation. Under no circumstances may the ID be used for test or real transfers!"

I do not want to get you guys in trouble after all. The documentation said that for testing purposes an ID with a leading 0 should be used. I could disable the leading 0 check if the `APP_ENV` is set to `TEST` if you like. Alternatively i could at least use my TIN for the German Test if you like.

**Comment by AlexOlival** [comment]
> I do not want to get you guys in trouble after all. The documentation said that for testing purposes an ID with a leading 0 should be used. I could disable the leading 0 check if the `APP_ENV` is set to `TEST` if you like. Alternatively i could at least use my TIN for the German Test if you like.

We'll trust the algorithm and we'll trust you :smile: 
In that case you can go ahead and use test IDs which they allow - no need for edge cases

**Comment by bofalke** [comment]
So i can leave everything like it is? :D Sorry for the confusion

**Comment by AlexOlival** [comment]
> So i can leave everything like it is? :D Sorry for the confusion

Use invalid IDs that the German authorities allow - whatever is legal

**Comment by bofalke** [comment]
So i will use testing IDs with a leading 0. That would then cause the test to fail because valid TIN's must not start with a 0. If you like i can at least use my own TIN to have one valid ID in the test.

**Comment by AlexOlival** [comment]
> So i will use testing IDs with a leading 0. That would then cause the test to fail because valid TIN's must not start with a 0. If you like i can at least use my own TIN to have one valid ID in the test.

Ahhhh I finally understood what you meant... 
Well I'm not a fan of environment variables and workarounds for testing in packages.
I also find it weird that we're not allowed to use IDs when these don't actually encode any information (especially if you used a generator for them)

I say we keep it as it is. If someone asks us to remove them, we will :man_shrugging:

**Comment by bofalke** [comment]
Ok perfect :+1: So this is all good now? :D

**Review by AlexOlival — APPROVED** [review]

**Comment by AlexOlival** [comment]
All good from my end!
@JoaoFSCruz como take a looksie

**Comment by AlexOlival** [comment]
@all-contributors please add @bofalke for documentation

**Review by JoaoFSCruz — APPROVED** [review]
